### ISSUE TITLE
feat: implement GraphSAGE algorithm based on GeaFlow inference framework

### ISSUE THREAD

**Issue by qingwen220** [TuGraph-family/tugraph-analytics#677]
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Comment by kitalkuyo-gita** [comment]
Detailed development plan to implement GraphSAGE based on GeaFlow inference framework

## 1. Framework architecture analysis

### 1.1 Core mechanism of GeaFlow-Infer inference framework

GeaFlow-Infer is a Java-Python hybrid inference framework that implements model inference through the following mechanisms:

**Core components:**
- **InferContext**: The inference context on the Java side, responsible for managing the inference task life cycle
- **Python Inference Service**: Communicates with Java processes through shared memory queues
- **DataBridge**: Implement Java-Python data exchange

### 1.2 Integration of reasoning framework and graph computing

The graph calculation passed`IncGraphInferContext`Interface call reasoning capabilities:

### 1.3 Python-side reasoning session mechanism

## 2. GraphSAGE algorithm implementation plan

### 2.1 Implement path selection

GraphSAGE can be implemented in **two ways**:

#### **Option A: Pure Infer framework implementation (recommended for inference scenarios)**
- Suitable for online inference on trained GraphSAGE models
- Utilizing geaflow-infer’s Python model calling capability
- Graph sampling is completed on the Java side, and feature aggregation is completed on the Python side through the model.

#### **Option B: UDGA framework implementation (suitable for training scenarios)**
- based on`AlgorithmUserFunction`Interface implementation
- Implement graph sampling and aggregation logic entirely on the Java side
- Suitable for scenarios where GraphSAGE training needs to be integrated into graph computing

**This solution focuses on solution A**, because it is more in line with the original design intention of geaflow-infer.

## 3. Detailed implementation plan (Plan A: based on Infer framework)

### 3.1 Overall architecture design

```mermaid
graph TB
    A[Java: IncVertexCentricCompute] --> B[邻居采样]
    B --> C[构造特征向量]
    C --> D[IncGraphInferContext.infer]
    D --> E[共享内存队列]
    E --> F[Python: GraphSAGE模型]
    F --> G[Mean/LSTM/Pool Aggregator]
    G --> H[返回节点嵌入]
    H --> E
    E --> D
    D --> I[更新节点表示]
```

### 3.2 Java side implementation

#### 3.2.1 Create GraphSAGE calculation task

Refer to existing reasoning examples

**Key implementation points:**

1. **Inherit IncVertexCentricCompute to implement incremental graph calculation**
2. **Implement multi-hop neighbor sampling logic in the compute function**, refer to the KHop algorithm

3. **Use AlgorithmRuntimeContext to load neighbor edges**

4. **Call Python model through IncGraphInferContext**

#### 3.2.2 Configure the inference environment

**Required configuration items:**
- `INFER_ENV_ENABLE`: enable inference framework
- `INFER_ENV_USER_TRANSFORM_CLASSNAME`: Python inference class name
- `INFER_ENV_INIT_TIMEOUT_SEC`: Environment initialization timeout
- `INFER_ENV_CONDA_URL`: Conda installation source (optional)

### 3.3 Python side implementation

#### 3.3.1 Implement TransFormFunction

Refer to the example structure in the documentation

**Key points of GraphSAGE Python implementation:**

1. **Inherits TransFormFunction abstract class**
2. **exist`__init__`Load the pre-trained GraphSAGE model**
3. **accomplish`transform_pre`method**:
   - Receive node ID and neighbor characteristics
   - Perform GraphSAGE aggregation operations (mean/LSTM/pool)
   - Return node embedding vector
4. **accomplish`transform_post`Method**: Post-processing results

#### 3.3.2 Data exchange mechanism

The Python side performs data exchange through pickle serialization:

### 3.4 GraphSAGE specific implementation code structure

#### 3.4.1 Java side core logic

```java
// GraphSAGECompute.java
public class GraphSAGECompute extends IncVertexCentricCompute<K, V, E, M> {
    
    @Override
    public IncVertexCentricComputeFunction<K, V, E, M> getIncComputeFunction() {
        return new GraphSAGEComputeFunction();
    }
    
    public static class GraphSAGEComputeFunction implements 
            IncVertexCentricComputeFunction<K, V, E, M> {
        
        private IncGraphInferContext<List<Double>> inferContext;
        private int numSamples = 10;  // 每层采样邻居数
        private int numLayers = 2;     // GraphSAGE层数
        
        @Override
        public void init(IncGraphComputeContext<K, V, E, M> context) {
            this.inferContext = (IncGraphInferContext<List<Double>>) context;
        }
        
        @Override
        public void evolve(K vertexId, TemporaryGraph<K, V, E> temporaryGraph) {
            // 1. 采样邻居（多层采样）
            Map<Integer, List<K>> sampledNeighbors = sampleNeighbors(vertexId);
            
            // 2. 收集节点特征和邻居特征
            Object[] features = prepareFeatures(vertexId, sampledNeighbors);
            
            // 3. 调用Python模型进行推理
            List<Double> embedding = inferContext.infer(features);
            
            // 4. 更新节点表示
            updateVertex(vertexId, embedding);
        }
        
        private Map<Integer, List<K>> sampleNeighbors(K vertexId) {
            // 实现邻居采样逻辑（类似KHop的多跳遍历）
            // 支持固定大小采样或重要性采样
        }
    }
}
```

#### 3.4.2 Python side core logic

```python
# TransFormFunctionUDF.py
import torch
import torch.nn as nn
from typing import List

class GraphSAGETransFormFunction(TransFormFunction):
    def __init__(self):
        super().__init__(input_size=3)  # vertexId, features, neighbor_features
        self.device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
        self.load_model('graphsage_model.pt')
    
    def load_model(self, model_path: str):
        # 加载预训练的GraphSAGE模型
        self.model = GraphSAGEModel(
            input_dim=128,
            hidden_dim=256,
            output_dim=64,
            num_layers=2,
            aggregator_type='mean'
        ).to(self.device)
        self.model.load_state_dict(torch.load(model_path))
        self.model.eval()
    
    def transform_pre(self, *args):
        # args[0]: 节点ID
        # args[1]: 节点特征
        # args[2]: 邻居节点特征字典 {layer: [neighbor_features]}
        
        vertex_id = args[0]
        vertex_feature = torch.tensor(args[1], dtype=torch.float32).to(self.device)
        neighbor_features = self._parse_neighbor_features(args[2])
        
        # 执行GraphSAGE聚合
        with torch.no_grad():
            embedding = self.model(vertex_feature, neighbor_features)
        
        return embedding.cpu().numpy().tolist(), vertex_id
    
    def transform_post(self, res):
        return res[0]  # 返回嵌入向量
    
    def _parse_neighbor_features(self, neighbor_data):
        # 解析邻居特征，组织成模型需要的格式
        pass

class GraphSAGEModel(nn.Module):
    def __init__(self, input_dim, hidden_dim, output_dim, num_layers, aggregator_type):
        super().__init__()
        self.num_layers = num_layers
        self.aggregator_type = aggregator_type
        
        # 定义GraphSAGE层
        self.layers = nn.ModuleList()
        for i in range(num_layers):
            in_dim = input_dim if i == 0 else hidden_dim
            out_dim = output_dim if i == num_layers - 1 else hidden_dim
            self.layers.append(GraphSAGELayer(in_dim, out_dim, aggregator_type))
    
    def forward(self, node_features, neighbor_features_list):
        h = node_features
        for i, layer in enumerate(self.layers):
            h = layer(h, neighbor_features_list[i])
        return h

class GraphSAGELayer(nn.Module):
    def __init__(self, in_dim, out_dim, aggregator_type='mean'):
        super().__init__()
        self.aggregator_type = aggregator_type
        
        if aggregator_type == 'mean':
            self.aggregator = MeanAggregator(in_dim, out_dim)
        elif aggregator_type == 'lstm':
            self.aggregator = LSTMAggregator(in_dim, out_dim)
        elif aggregator_type == 'pool':
            self.aggregator = PoolAggregator(in_dim, out_dim)
    
    def forward(self, node_feature, neighbor_features):
        return self.aggregator(node_feature, neighbor_features)
```

#### 3.4.3 requirements.txt

```text
--index-url https://pypi.tuna.tsinghua.edu.cn/simple
torch>=1.12.0
torch-geometric>=2.3.0
numpy>=1.21.0
scikit-learn>=1.0.0
```

### 3.5 Neighbor sampling strategy

#### 3.5.1 Fixed size sampling (recommended)
- Fixed sampling K neighbors per layer
- Repeat sampling or padding when there are less than K
- Simple implementation and stable performance

#### 3.5.2 Importance Sampling
- Weighted sampling based on edge weight or degree
- More suitable for heterogeneous graphs and weighted graphs

#### 3.5.3 Multi-layer sampling example

Refer to the neighbor traversal of CommonNeighbors:
## 4. Production environment deployment considerations

### 4.1 Performance optimization

1. **Batch Reasoning**:
   - Call models in batches after accumulating multiple nodes
   - Leveraging GPUs to accelerate batch computation
   - exist`transform_pre`Processing batch data

2. **Memory Management**:
   - Control the number of neighbor samples to avoid memory explosion
   - Use an iterator instead of loading all edges:

3. **Model Optimization**:
   - Compile the model using TorchScript
   - Consider quantization and pruning
   - Accelerate with ONNX Runtime

### 4.2 Environment configuration

The inference environment is managed through conda:

**Production environment recommendations:**
- Pre-built conda environment including all dependencies
- Set a reasonable timeout (large models require longer initialization time)
- Configure appropriate LD_LIBRARY_PATH

### 4.3 Fault tolerance and monitoring

1. **Exception handling**:

2. **Process Monitoring**: The Python process will automatically detect the status of the parent process
3. **Logging**: Make full use of logs to track the reasoning process

### 4.4 Data consistency

- **Feature synchronization**: Ensure that the feature formats on the Java side and the Python side are consistent
- **ID Mapping**: Maintain global consistency of node IDs
- **Version Control**: Correspondence between model files and code versions

## 5. Development steps

### 5.1 Phase 1: Environment setup (1-2 days)
1. Compile the GeaFlow source code and ensure that the geaflow-infer module is available
2. Prepare Python environment and GraphSAGE model dependencies
3. Verify the basic functions of the reasoning framework

### 5.2 Phase 2: Prototype development (3-5 days)
1. Implement Java side neighbor sampling logic
2. Implement Python-side GraphSAGE model loading and inference
3. Implement data serialization and feature transfer
4. Validate functionality on small-scale data sets

### 5.3 The third stage: optimization and testing (3-5 days)
1. Implement batch inference optimization
2. Add exception handling and logging
3. Performance benchmarks
4. Large-scale data set validation

### 5.4 The fourth stage: production (2-3 days)
1. Write complete documentation
2. Containerized deployment
3. Monitoring alarm configuration
4. Production environment trial run

## 6. Code Organization Structure

```
GraphSAGE-GeaFlow/
├── src/main/java/
│   └── org/example/graphsage/
│       ├── GraphSAGECompute.java          # 主计算逻辑
│       ├── GraphSAGEComputeFunction.java  # 计算函数
│       ├── NeighborSampler.java           # 邻居采样器
│       └── FeatureCollector.java          # 特征收集器
├── src/main/resources/
│   └── TransFormFunctionUDF.py            # Python推理逻辑
│       ├── graphsage_model.py             # GraphSAGE模型定义
│       └── aggregators.py                 # 聚合器实现
├── requirements.txt                        # Python依赖
├── models/
│   └── graphsage_model.pt                 # 预训练模型
└── pom.xml                                # Maven配置
```
